modules/etcd/etcd.go

```diff
@@ -0,0 +1,266 @@
+package etcd
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/testcontainers/testcontainers-go"
+	tcexec "github.com/testcontainers/testcontainers-go/exec"
+	tcnetwork "github.com/testcontainers/testcontainers-go/network"
+)
+
+const (
+	ClientPort          = "2379"
+	PeerPort            = "2380"
+	DataDir             = "/data.etcd"
+	DefaultClusterToken = "mys3cr3ttok3n"
+)
+
+// EtcdContainer represents the etcd container type used in the module
+type EtcdContainer struct {
+	testcontainers.Container
+	ClusterToken string
+	Nodes        []*EtcdContainer
+}
+
+// Run creates an instance of the etcd container type
+func Run(ctx context.Context, img string, opts ...testcontainers.ContainerCustomizer) (*EtcdContainer, error) {
+	req := testcontainers.ContainerRequest{
+		Image: img,
+		Cmd:   []string{},
+	}
+
+	genericContainerReq := testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	}
+
+	settings := defaultOptions()
+	for _, opt := range opts {
+		if apply, ok := opt.(Option); ok {
+			apply(&settings)
+		}
+		if err := opt.Customize(&genericContainerReq); err != nil {
+			return nil, err
+		}
+	}
+
+	// configure data dir
+	if settings.MountDataDir {
+		genericContainerReq.LifecycleHooks = append(genericContainerReq.LifecycleHooks, testcontainers.ContainerLifecycleHooks{
+			PostStarts: []testcontainers.ContainerHook{
+				func(ctx context.Context, c testcontainers.Container) error {
+					_, _, err := c.Exec(ctx, []string{"chmod", "o+rwx", "-R", DataDir}, tcexec.Multiplexed())
+					if err != nil {
+						return fmt.Errorf("chmod etcd data dir: %w", err)
+					}
+
+					return nil
+				},
+			},
+		})
+	}
+
+	clusterOpts, err := configureCluster(ctx, &settings, opts)
+	if err != nil {
+		return nil, fmt.Errorf("configure cluster: %w", err)
+	}
+
+	// configure CMD with the nodes
+	genericContainerReq.Cmd = configureCMD(settings)
+
+	if settings.clusterNetwork != nil {
+		// apply the network to the current node
+		err := tcnetwork.WithNetwork([]string{settings.Nodes[settings.currentNode]}, settings.clusterNetwork)(&genericContainerReq)
+		if err != nil {
+			return nil, fmt.Errorf("with network: %w", err)
+		}
+	}
+
+	container, err := testcontainers.GenericContainer(ctx, genericContainerReq)
+	var c *EtcdContainer
+	if container != nil {
+		c = &EtcdContainer{Container: container}
+	}
+
+	if err != nil {
+		return c, fmt.Errorf("generic container: %w", err)
+	}
+
+	// only the first node creates the cluster
+	if settings.currentNode == 0 {
+		for i := 1; i < len(settings.Nodes); i++ {
+			clusterOpts = append(clusterOpts, withCurrentNode(i)) // move to the next node
+			clusterNode, err := Run(ctx, req.Image, clusterOpts...)
```

Comment:
bug: while this will likely result in the correct behaviour I don't think having multiple withCurrentNode on clustOptions is the intent.
```suggestion
			clusterNode, err := Run(ctx, req.Image, append(clusterOpts, withCurrentNode(i))...)
```